public/app/features/alerting/AlertRuleList.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { PureComponent } from 'react';
+import { hot } from 'react-hot-loader';
+import { connect } from 'react-redux';
+import PageHeader from 'app/core/components/PageHeader/PageHeader';
+import AlertRuleItem from './AlertRuleItem';
+import appEvents from 'app/core/app_events';
+import { updateLocation } from 'app/core/actions';
+import { getNavModel } from 'app/core/selectors/navModel';
+import { NavModel, StoreState, AlertRule } from 'app/types';
+import { getAlertRulesAsync, setSearchQuery } from './state/actions';
+import { getAlertRuleItems, getSearchQuery } from './state/selectors';
+
+export interface Props {
+  navModel: NavModel;
+  alertRules: AlertRule[];
+  updateLocation: typeof updateLocation;
+  getAlertRulesAsync: typeof getAlertRulesAsync;
+  setSearchQuery: typeof setSearchQuery;
+  stateFilter: string;
+  search: string;
+}
+
+export class AlertRuleList extends PureComponent<Props, any> {
+  stateFilters = [
+    { text: 'All', value: 'all' },
+    { text: 'OK', value: 'ok' },
+    { text: 'Not OK', value: 'not_ok' },
+    { text: 'Alerting', value: 'alerting' },
+    { text: 'No Data', value: 'no_data' },
+    { text: 'Paused', value: 'paused' },
+  ];
+
+  componentDidMount() {
+    this.fetchRules();
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.stateFilter !== this.props.stateFilter) {
+      this.fetchRules();
+    }
+  }
+
+  async fetchRules() {
+    await this.props.getAlertRulesAsync({ state: this.getStateFilter() });
+  }
+
+  getStateFilter(): string {
+    const { stateFilter } = this.props;
+    if (stateFilter) {
+      return stateFilter.toString();
+    }
+    return 'all';
+  }
+
+  onStateFilterChanged = event => {
+    this.props.updateLocation({
+      query: { state: event.target.value },
+    });
+  };
+
+  onOpenHowTo = () => {
+    appEvents.emit('show-modal', {
+      src: 'public/app/features/alerting/partials/alert_howto.html',
+      modalClass: 'confirm-modal',
+      model: {},
+    });
+  };
+
+  onSearchQueryChange = event => {
+    const { value } = event.target;
+    this.props.setSearchQuery(value);
+  };
+
+  alertStateFilterOption = ({ text, value }) => {
+    return (
+      <option key={value} value={value}>
+        {text}
+      </option>
+    );
+  };
+
+  render() {
+    const { navModel, alertRules, search } = this.props;
+
+    return (
+      <div>
+        <PageHeader model={navModel} />
```

Comment:
Seems like this component will be used in many "containers". Would it be possible to create a higher level page container that basically would have a "layout" for rendering the header and connect to redux store etc and then take page in question as prop - in this case AlertRuleList. 

I remember back in the days when I looked at react router you could declare routes using "layouts", like [this](https://simonsmith.io/reusing-layouts-in-react-router-4/). 

Understand that we cannot use react router until we got rid of angular routing so what I'm asking for is basically some layout component that you can wrap your component in, for example AlertRuleList. Then AlertRuleList doesn't need to know/care about the PageHeader component.

Reply:
yes, maybe once more is in react & redux, the PageHeader is used by angular & react pages (redux & mobx) right now. Once we know how dynamic pages will work (like team edit page, etc) that require more interaction between container (loading team details) -> nav Index -> build new navModel, then it might be quite easy to move this higher up. But since we cant use react router yet it might still be problematic

Reply:
For now I was thinking that you could do something like this:

Pseudo code!

```javascript
// SomePageLayout
render() {
  return (
      <div>
        <PageHeader model={navModel} />
       {{children}}
     </div>
  }
}

// AlertRulePage
render() {
  return <SomePageLayout children=<AlertRuleList/> />
}
```

Reply:
Yes, been thinking about trying something like that, won't you need to pass in navModel to SomePageLayout there as well there? 

> Instead of render in AlertRulePage maybe you can just create a wrapper that you provide to the connect call?

What do you mean?

Reply:
Exactly how you solve that it's out of my knowledge :smile: But yes you need to pass navModel. 

> What do you mean?

I was thinking that maybe you could compose an AlertRuleList component using something like this pseudo code:

```javascript
const wrapped = withSomePageLayout(AlertRuleList);
...
export default hot(module)(connect(mapStateToProps, mapDispatchToProps)(wrapped));
```

Is the correct wording a higher order component maybe?

Reply:
could work, but you would also have to send in navModel, which for dynamic pages is dynamic, so not exactly sure how that will work in this case. but having a higher order component handle common page layout is interesting. But all the page layout stuff is in PageHeader anyway so not that big a win yet, but if we had more common page layout I think we definitely need this. (for example when sidemenu and footer is react container as well and grafana app is react from the root)